core/src/HIP/Kokkos_HIP_ReduceScan.hpp

```diff
@@ -0,0 +1,209 @@
+      ::Kokkos::Experimental::HIP::size_type const block_count,
+      ::Kokkos::Experimental::HIP::size_type* const shared_data,
+      ::Kokkos::Experimental::HIP::size_type* const global_data,
+      ::Kokkos::Experimental::HIP::size_type* const global_flags) {
+    Scalar* const global_team_buffer_element =
+        reinterpret_cast<Scalar*>(global_data);
+    Scalar* const my_global_team_buffer_element =
+        global_team_buffer_element + hipBlockIdx_x;
+    Scalar* shared_team_buffer_elements =
+        reinterpret_cast<Scalar*>(shared_data);
+    Scalar value        = shared_team_buffer_elements[hipThreadIdx_y];
+    int shared_elements = (hipBlockDim_x * hipBlockDim_y) /
+                          ::Kokkos::Experimental::Impl::HIPTraits::WarpSize;
+    int global_elements = block_count;
+    __syncthreads();
+
+    // Do the scalar reduction inside each block
+    scalar_intra_block_reduction(functor, value, true,
+                                 my_global_team_buffer_element, shared_elements,
+                                 shared_team_buffer_elements);
+    __syncthreads();
+
+    // Use the last block that is done to do the do the reduction across the
+    // block
+    __shared__ unsigned int num_teams_done;
+    if (hipThreadIdx_x + hipThreadIdx_y == 0) {
+      __threadfence();
+      num_teams_done = Kokkos::atomic_fetch_add(global_flags, 1) + 1;
+    }
+    bool is_last_block = false;
+    // TODO HIP does not support syncthreads_or. That's why we need to make
+    // num_teams_done __shared__
+    // if (__syncthreads_or(num_teams_done == hipGridDim_x)) {*/
+    __syncthreads();
+    if (num_teams_done == hipGridDim_x) {
+      is_last_block = true;
+      *global_flags = 0;
+      ValueInit::init(functor, &value);
+      for (int i = hipThreadIdx_y * hipBlockDim_x + hipThreadIdx_x;
+           i < global_elements; i += hipBlockDim_x * hipBlockDim_y) {
+        ValueJoin::join(functor, &value, &global_team_buffer_element[i]);
+      }
+      scalar_intra_block_reduction(
+          functor, value, false,
+          shared_team_buffer_elements + (hipBlockDim_y - 1), shared_elements,
+          shared_team_buffer_elements);
+    }
+
+    return is_last_block;
+  }
+};
+
+template <bool DoScan, typename FunctorType, typename ArgTag>
+__device__ bool hip_single_inter_block_reduce_scan(
+    FunctorType const& functor,
+    ::Kokkos::Experimental::HIP::size_type const block_id,
+    ::Kokkos::Experimental::HIP::size_type const block_count,
+    ::Kokkos::Experimental::HIP::size_type* const shared_data,
+    ::Kokkos::Experimental::HIP::size_type* const global_data,
+    ::Kokkos::Experimental::HIP::size_type* const global_flags) {
+  using ValueTraits = FunctorValueTraits<FunctorType, ArgTag>;
+  if (!DoScan && ValueTraits::StaticValueSize)
+    // TODO For now we don't use shuffle
+    // return Kokkos::Impl::HIPReductionsFunctor<
+    //    FunctorType, ArgTag, false, (ValueTraits::StaticValueSize > 16)>::
```

Comment:
What was in that specialization that we cannot implement yet?

Reply:
Nothing, we can implement it whenever we want. I just wanted to have something that works and then add more code.